### PR TITLE
Enable Blockwise Int4 quantized linear layer

### PR DESCRIPTION
**Pull request by lsy323**
### User journey

Different quantize methods can be configured by `--quantize_type="int8_per_channel"/"int4_per_channel"/"int8_blockwise"/"int4_blockwise"` when running `run_server.py`, `run_offline.py` and `run_interactive.py`. (`README` is also updated accordingly)

- Quantize bf16 checkpoint with int4 blockwise quantization for linear layers. (Embedding layers will be kept in int8 per-channel quant if int4 or blockwise quant is enabled)
```
python -m convert_checkpoints --input_checkpoint_dir=$input_ckpt_dir --output_checkpoint_dir=$output_ckpt_dir \
    --quantize=True \
    --quantize_type="int4_blockwise"
```
- Run `run_interactive`, similar for `run_offline.py` and `run_server.py`
```
python run_interactive.py --model_name='llama-2' --size=7b --batch_size=2 --max_cache_length=2048 \
    --quantize_weights=True --quantize_kv_cache=True \
    --quantize_type="int4_blockwise" \
    --checkpoint_path=$output_ckpt_dir \
    --tokenizer_path=$tokenizer_path
```

### Quantization config workflow
Quantization configs are stored in `QuantizationConfig` dataclass, Environment stores a `QuantizationConfig` instance. Model initiate quantized layers based on the quantization config in environment.

### Int4 weight loading workflow:
1. There is no `torch.int4`, so `convert_checkpoints` will store the in4 weights in int8 container
2. When the jax state_dict is extracted from checkpoint in `engine.py`, we cast the int8 tensors to int4 JAX tensors.

### New quantization support
1. Added {int8, int4} x {per_channel, blockwise} quantized linear layers.
2. Added asymmetric quant support to quant/dequant function and quantized layers, but it's not exposed to cmd config, for experimental purpose.

Changes:
- Add quantization cmd configs.
- Add support of int4 and blockwise quantization in `convert_checkpoint.py`, move quantization map to model class.
- Implementation of blockwise quantized linear layer. Explored different implementations and only the `einsum without flatten operands` works the best, the others are still kept in case they become useful in the future.
- Int4 weight loading in `engine.py`
- Refactor the weight loading logic in `engine.py`
- Update quantize and dequantize function (Enhancement accuracy by adding `eps`, `round`, blockwise and asymmetric quant support)

Test:
Correctness
- Added unit tests for quantize/dequantize function, quantized linear accuracy, blockwise quant sharding.
- `run_interactive` to verify the int4 blockwise quant generates meaning tokens
Latency
- Summarize in an internal doc, it's not always outperforming int8 per-channel, but int4 blockwise matmul kernel is faster.

### CODE REVIEW

**Comment by lsy323** [comment]
This needs https://github.com/pytorch/xla/pull/7071 to land on PyTorch/XLA side, will udpate the pin once it lands

File: benchmarks/run_offline.py [hunk 1]
```diff
@@ -62,13 +65,18 @@
 _QUANTIZE_WEIGHTS = flags.DEFINE_bool(
     "quantize_weights", False, "weight quantization"
 )
+_QUANTIZE_NUM_BITS_WEIGHTS = flags.DEFINE_integer(
```
Comment (qihqi) [inline benchmarks/run_offline.py:68]: combine these into `quantize_type` which can take values of int8, int4, int4_block

File: jetstream_pt/layers.py [hunk 2]
```diff
@@ -59,15 +81,275 @@ def __init__(self, in_features, out_features, bias=None, device=None):
...
   def forward(self, inputs):
-    return F.linear(inputs, self.weight) * self.weight_scaler
+    if self.is_symmetric:
+      return torch.mul(F.linear(inputs, self.weight), self.weight_scaler)
+    else:
+      out = torch.mul(F.linear(inputs, self.weight), self.weight_scaler)
+      zp_out = torch.einsum("...c,z->...z", inputs, self.zero_point)
+      return out - zp_out
+
+    if self.run_fake_quantize:
```
Comment (qihqi) [inline jetstream_pt/layers.py:128]: does this has an else branch?

Comment (lsy323) [reply jetstream_pt/layers.py:128]: I can do 
```
if self.run_fake_quantize:
 ...
else:
  if self.is_symmetric:
    ...
  else:
    ...
```

Comment (qihqi) [reply jetstream_pt/layers.py:128]: sure.

File: jetstream_pt/layers.py [hunk 3]
```diff
@@ -59,15 +81,275 @@ def __init__(self, in_features, out_features, bias=None, device=None):
...
+    """
+    self.weight, self.weight_scaler, self.zero_point = load_q_weight_helper(
+        w_q, scale, zp, block_size=-1
+    )
+
+  def quantize_weight_from_nn_linear(self, weight):
+    assert weight.dim() == 2, "Expect 2D weight from torch.nn.Linear."
+    assert weight.shape == (
+        self.out_features,
+        self.in_features,
```
Comment (FanhaiLu1) [inline jetstream_pt/layers.py:111]: nit: add expected weight shape in assert error message to help engineer figure out what are correct shape when it's wrong.

File: jetstream_pt/layers.py [hunk 4]
```diff
@@ -59,15 +81,275 @@ def __init__(self, in_features, out_features, bias=None, device=None):
...
+    ), f"Unexpected weight shape ({self.out_features}, {self.in_features})."
+    w_q, scale, zp = quantize_tensor(
+        weight, (1,), self.n_bit, self.is_symmetric, block_size=-1
+    )
+    w_dq = dequantize_tensor(w_q, scale, zp)
+    print("check qweight cosine dist: ", _calc_cosine_dist(weight, w_dq))
+    self._load_quantized_weights(w_q, scale, zp)
```
Comment (FanhaiLu1) [inline jetstream_pt/layers.py:118]: remove print?

File: jetstream_pt/layers.py [hunk 5]
```diff
@@ -59,15 +81,275 @@ def __init__(self, in_features, out_features, bias=None, device=None):
...
 
   def forward(self, inputs):
-    return F.linear(inputs, self.weight) * self.weight_scaler
+    if self.is_symmetric:
+      return torch.mul(F.linear(inputs, self.weight), self.weight_scaler)
+    else:
```
Comment (FanhaiLu1) [inline jetstream_pt/layers.py:123]: Do we already support asymmetric quantization?

File: benchmarks/run_offline.py (hunk 1 above)
Comment (lsy323) [reply benchmarks/run_offline.py]: Agree, done

File: jetstream_pt/layers.py (hunk 2 above)
Comment (lsy323) [reply jetstream_pt/layers.py:127]: done

File: jetstream_pt/layers.py (hunk 3 above)
Comment (lsy323) [reply jetstream_pt/layers.py:111]: Makes sense! Done

File: jetstream_pt/layers.py (hunk 4 above)
Comment (lsy323) [reply jetstream_pt/layers.py:117]: done

File: jetstream_pt/layers.py (hunk 5 above)
Comment (lsy323) [reply jetstream_pt/layers.py:122]: Didn't expose it to cmd yet. numerical correctness is verified in tests.

File: benchmarks/run_offline.py [hunk 6]
```diff
@@ -134,9 +98,13 @@ def main(argv):
     )
     sampled_tokens_list.append(sampled_tokens)
 
+  profiling_output = FLAGS.profiling_output
   print("======= decode starting ===")
   dec_times = []
   for i in range(10):
+    if i == 7:
```
Comment (FanhaiLu1) [inline benchmarks/run_offline.py:105]: Nit: merge two if in one line.

File: jetstream_pt/third_party/llama/model_exportable.py [hunk 7]
```diff
@@ -199,3 +204,44 @@ def forward(
       h = self.norm(h)
       output = self.output(h).float()
     return output
+
+  @staticmethod
+  def get_quantized_linear_weight_to_scaler_map():
```
Comment (FanhaiLu1) [inline jetstream_pt/third_party/llama/model_exportable.py:209]: Would you consider to move these 3 static method to a quantization until class?

File: jetstream_pt/layers.py [hunk 8]
```diff
@@ -59,15 +81,280 @@ def __init__(self, in_features, out_features, bias=None, device=None):
...
+    w_q, scale, zp = quantize_tensor(
+        weight, (1,), self.n_bit, self.is_symmetric, self.block_size
+    )
+    w_dq = dequantize_tensor(w_q, scale, zp)
+    print("check qweight cosine dist: ", _calc_cosine_dist(weight, w_dq))
+    # breakpoint()
+    self._load_quantized_weights(w_q, scale, zp)
+
+  @staticmethod
+  def blockwise_jax_kernel(j_inputs, j_weight, j_weight_scaler, j_zero_point):
```
Comment (FanhaiLu1) [inline jetstream_pt/layers.py:218]: Is there any reason to have "j" as prefix of parameter name?

File: jetstream_pt/third_party/llama/model_exportable.py (hunk 7 above)
Comment (lsy323) [reply jetstream_pt/third_party/llama/model_exportable.py:209]: Actually I intended to keep them as static methods of a model class, since they are 'attributes' of the model and are model specific. (e.g. gemma may have different maps)

File: benchmarks/run_offline.py (hunk 6 above)
Comment (lsy323) [reply benchmarks/run_offline.py:105]: Sure, let me update

File: jetstream_pt/layers.py (hunk 8 above)
Comment (lsy323) [reply jetstream_pt/layers.py:218]: Yes, this is passed into `torchjax.call_jax`, the parameters are jax tensors, so I name them with a prefix j.

File: benchmarks/run_offline.py (hunk 6 above)
Comment (lsy323) [reply benchmarks/run_offline.py]: Done, thanks for catching it.

File: jetstream_pt/layers.py (hunk 8 above)
Comment (FanhaiLu1) [reply jetstream_pt/layers.py:218]: Thanks for sharing it! I feel it would be more readable if removing the prefix j and adding a comment to mention all these are jax tensor under this function.

**Review by FanhaiLu1 — APPROVED** [review]

**Review by qihqi — APPROVED** [review]

File: jetstream_pt/layers.py (hunk 8 above)
Comment (lsy323) [reply jetstream_pt/layers.py]: Updated
